File: ios/FitnessGeolocation/ScheduleManager.swift
import Foundation

/// Cron-like schedule windows — Transistorsoft-compatible format.
/// Examples: "1-7 09:00-17:00", "1-5 08:00-18:00 geofence"
final class ScheduleManager {
  static let shared = ScheduleManager()

  enum TrackingMode: String {
    case location
    case geofence
  }

  struct Window {
    let days: Set<Int>       // 1=Sunday … 7=Saturday (US)
    let onMinutes: Int       // minutes from midnight
    let offMinutes: Int
    let mode: TrackingMode
    var triggered = false
  }

  private(set) var windows: [Window] = []
  private(set) var isEnabled = false
  var onScheduleChange: ((Bool, TrackingMode) -> Void)?

  private init() {}

  func configure(records: [String]) {
    windows = records.compactMap { parse($0) }
  }

  func start() {
    isEnabled = true
    evaluate()
  }

  func stop() {
    isEnabled = false
    windows.indices.forEach { windows[$0].triggered = false }
  }

  /// Call on location tick, app resume, and background fetch.
  func evaluate(now: Date = Date()) {
    guard isEnabled, !windows.isEmpty else { return }
    let cal = Calendar.current
    let weekday = cal.component(.weekday, from: now) // 1=Sun
    let minutes = cal.component(.hour, from: now) * 60 + cal.component(.minute, from: now)

    for i in windows.indices {
      var w = windows[i]
      guard w.days.contains(weekday) else { continue }
      let inWindow: Bool
      if w.onMinutes <= w.offMinutes {
        inWindow = minutes >= w.onMinutes && minutes < w.offMinutes
      } else {
        // overnight window e.g. 22:00-06:00
        inWindow = minutes >= w.onMinutes || minutes < w.offMinutes
      }
      if inWindow && !w.triggered {
        w.triggered = true
        windows[i] = w
        onScheduleChange?(true, w.mode)
      } else if !inWindow && w.triggered {
        w.triggered = false
        windows[i] = w
        onScheduleChange?(false, w.mode)
      }
    }
  }

  func stateDict() -> [String: Any] {
    ["schedulerEnabled": isEnabled, "scheduleCount": windows.count]
  }

  // MARK: - Parser

  private func parse(_ record: String) -> Window? {
    let parts = record.trimmingCharacters(in: .whitespaces).split(separator: " ", omittingEmptySubsequences: true)
    guard parts.count >= 2 else { return nil }
    let dayPart = String(parts[0])
    let timePart = String(parts[1])
    let mode: TrackingMode = parts.count >= 3 && parts[2].lowercased() == "geofence" ? .geofence : .location

    guard let days = parseDays(dayPart),
          let (on, off) = parseTimeRange(timePart) else { return nil }
    return Window(days: days, onMinutes: on, offMinutes: off, mode: mode)
  }

  private func parseDays(_ s: String) -> Set<Int>? {
    if s.contains("-") {
      let bounds = s.split(separator: "-")
      guard bounds.count == 2,
            let lo = Int(bounds[0]), let hi = Int(bounds[1]) else { return nil }
      return Set(lo...hi)
    }
    if let d = Int(s) { return [d] }
    return nil
  }

  private func parseTimeRange(_ s: String) -> (Int, Int)? {
    let bounds = s.split(separator: "-")
    guard bounds.count == 2,
          let on = parseTime(String(bounds[0])),
          let off = parseTime(String(bounds[1])) else { return nil }
    return (on, off)
  }

  private func parseTime(_ s: String) -> Int? {
    let p = s.split(separator: ":")
    guard p.count == 2, let h = Int(p[0]), let m = Int(p[1]) else { return nil }
    return h * 60 + m
  }
}

File: android/src/main/java/com/fitnessgeolocation/ScheduleManager.kt
package com.fitnessgeolocation

import java.util.Calendar

/** Cron-like schedule windows — Transistorsoft-compatible format. */
class ScheduleManager {
  enum class TrackingMode { location, geofence }

  data class Window(
    val days: Set<Int>,
    val onMinutes: Int,
    val offMinutes: Int,
    val mode: TrackingMode,
    var triggered: Boolean = false,
  )

  private val windows = mutableListOf<Window>()
  var isEnabled = false
    private set
  var onScheduleChange: ((Boolean, TrackingMode) -> Unit)? = null

  fun configure(records: List<String>) {
    windows.clear()
    records.forEach { parse(it)?.let { w -> windows.add(w) } }
  }

  fun start() {
    isEnabled = true
    evaluate()
  }

  fun stop() {
    isEnabled = false
    windows.forEach { it.triggered = false }
  }

  fun evaluate(now: Calendar = Calendar.getInstance()) {
    if (!isEnabled || windows.isEmpty()) return
    val weekday = now.get(Calendar.DAY_OF_WEEK) // 1=Sunday
    val minutes = now.get(Calendar.HOUR_OF_DAY) * 60 + now.get(Calendar.MINUTE)

    windows.forEachIndexed { i, w ->
      if (!w.days.contains(weekday)) return@forEachIndexed
      val inWindow = if (w.onMinutes <= w.offMinutes) {
        minutes in w.onMinutes until w.offMinutes
      } else {
        minutes >= w.onMinutes || minutes < w.offMinutes
      }
      if (inWindow && !w.triggered) {
        windows[i] = w.copy(triggered = true)
        onScheduleChange?.invoke(true, w.mode)
      } else if (!inWindow && w.triggered) {
        windows[i] = w.copy(triggered = false)
        onScheduleChange?.invoke(false, w.mode)
      }
    }
  }

  fun stateMap(): Map<String, Any> = mapOf(
    "schedulerEnabled" to isEnabled,
    "scheduleCount" to windows.size,
  )

  private fun parse(record: String): Window? {
    val parts = record.trim().split(" ").filter { it.isNotEmpty() }
    if (parts.size < 2) return null
    val days = parseDays(parts[0]) ?: return null
    val (on, off) = parseTimeRange(parts[1]) ?: return null
    val mode = if (parts.size >= 3 && parts[2].equals("geofence", true)) TrackingMode.geofence else TrackingMode.location
    return Window(days, on, off, mode)
  }

  private fun parseDays(s: String): Set<Int>? {
    if ("-" in s) {
      val b = s.split("-")
      if (b.size != 2) return null
      val lo = b[0].toIntOrNull() ?: return null
      val hi = b[1].toIntOrNull() ?: return null
      return (lo..hi).toSet()
    }
    return s.toIntOrNull()?.let { setOf(it) }
  }

  private fun parseTimeRange(s: String): Pair<Int, Int>? {
    val b = s.split("-")
    if (b.size != 2) return null
    val on = parseTime(b[0]) ?: return null
    val off = parseTime(b[1]) ?: return null
    return on to off
  }

  private fun parseTime(s: String): Int? {
    val p = s.split(":")
    if (p.size != 2) return null
    val h = p[0].toIntOrNull() ?: return null
    val m = p[1].toIntOrNull() ?: return null
    return h * 60 + m
  }
}
